### PR TITLE
Move `TranslatorSv2` lib code out of `main.rs`

### PR DESCRIPTION
**Pull request by jbesraa**
Related to #1093

### CODE REVIEW

File: roles/translator/src/lib/proxy_config.rs
Comment (Shourya742) [inline roles/translator/src/lib/proxy_config.rs:53]: Since these structs are open to anyone starting a translator and are used for the initial proxy setup, it’s important to make sure they're easy to use and less error-prone. These structs have a lot of fields and can get pretty lengthy, so I suggest using the Builder design pattern to simplify things. This approach will make initialization less prone to mistakes and more straightforward. Here’s a quick example:

```
#[derive(Debug, Deserialize, Clone)]
pub struct ProxyConfig {
    pub upstream_address: String,
    pub upstream_port: u16,
    pub upstream_authority_pubkey: Secp256k1PublicKey,
    pub downstream_address: String,
    pub downstream_port: u16,
    pub max_supported_version: u16,
    pub min_supported_version: u16,
    pub min_extranonce2_size: u16,
    pub downstream_difficulty_config: DownstreamDifficultyConfig,
    pub upstream_difficulty_config: UpstreamDifficultyConfig,
}

pub struct ProxyConfigBuilder {
    upstream: Option<TranslatorProxyUpstream>,
    downstream: Option<TranslatorProxyDownstream>,
    max_supported_version: Option<u16>,
    min_supported_version: Option<u16>,
    min_extranonce2_size: Option<u16>,
}

impl ProxyConfigBuilder {
    pub fn builder() -> Self {
        Self {
            upstream: None,
            downstream: None,
            max_supported_version: None,
            min_supported_version: None,
            min_extranonce2_size: None,
        }
    }

    pub fn upstream(mut self, upstream: TranslatorProxyUpstream) -> Self {
        self.upstream = Some(upstream);
        self
    }

    pub fn downstream(mut self, downstream: TranslatorProxyDownstream) -> Self {
        self.downstream = Some(downstream);
        self
    }

    pub fn max_supported_version(mut self, version: u16) -> Self {
        self.max_supported_version = Some(version);
        self
    }

    pub fn min_supported_version(mut self, version: u16) -> Self {
        self.min_supported_version = Some(version);
        self
    }

    pub fn min_extranonce2_size(mut self, size: u16) -> Self {
        self.min_extranonce2_size = Some(size);
        self
    }

    pub fn build(self) -> Result<ProxyConfig, &'static str> {
        Ok(ProxyConfig {
            upstream_address: self.upstream.ok_or("Upstream not set")?.address,
            upstream_port: self.upstream.ok_or("Upstream not set")?.port,
            upstream_authority_pubkey: self.upstream.ok_or("Upstream not set")?.authority_pubkey,
            downstream_address: self.downstream.ok_or("Downstream not set")?.address,
            downstream_port: self.downstream.ok_or("Downstream not set")?.port,
            max_supported_version: self.max_supported_version.ok_or("Max version not set")?,
            min_supported_version: self.min_supported_version.ok_or("Min version not set")?,
            min_extranonce2_size: self.min_extranonce2_size.ok_or("Min extranonce size not set")?,
            downstream_difficulty_config: self.downstream.ok_or("Downstream not set")?.difficulty_config,
            upstream_difficulty_config: self.upstream.ok_or("Upstream not set")?.difficulty_config,
        })
    }
}
```
So, the public API would look something like this:
```
let proxy_config = ProxyConfigBuilder::builder().upstream(upstream).downstream(downstream).max_supported_version(2).min_supported_version(1).min_extranonce2_size(8).build()
```

This makes it easier to set up and less likely to cause errors.

cc: @plebhash @GitGab19 @jbesraa

**Review by Shourya742 — CHANGES_REQUESTED** [review]
Concept ACK

File: roles/translator/src/lib/proxy_config.rs
Comment (GitGab19) [reply roles/translator/src/lib/proxy_config.rs:53]: I don't see any cons in doing that. I just think that if we are going to use a specific pattern, it should be the same everywhere

Comment (marathon-gary) [reply roles/translator/src/lib/proxy_config.rs:53]: If this change is desired (I like builder patterns, personally), it should be a separate PR and issue, imho. Would make for a `good first issue`

File: roles/translator/src/lib/proxy_config.rs
Comment (marathon-gary) [inline roles/translator/src/lib/proxy_config.rs:32]: It feels like there could be additional limitation to the extranonce and version fields here. Definitely outside the scope of this PR but worth mentioning for discussion.

**Review by marathon-gary — COMMENTED** [review]
lgtm pending updating with base branch

File: roles/translator/src/lib/proxy_config.rs
Comment (GitGab19) [reply roles/translator/src/lib/proxy_config.rs:53]: > If this change is desired (I like builder patterns, personally), it should be a separate PR and issue, imho. Would make for a `good first issue`

Completely agree  💯 
Maybe we could push this one which could be used as an example for the `good first issue`(s)

**Comment by Shourya742** [comment]
I tested the branch locally, and the logs for the translator look fine. However, I noticed quite a few warnings. I believe some of the error cases are no longer being triggered. If possible, could you remove those and resolve the other warnings as well? Everything else looks good.

<details>
<summary>
Warning Logs Translator
</summary>
<code>

fields `extranonce` and `extranonce2_len` are never read
  --> translator/src/lib/downstream_sv1/mod.rs:27:9
   |
24 | pub struct SubmitShareWithChannelId {
   |            ------------------------ fields in this struct
...
27 |     pub extranonce: Vec<u8>,
   |         ^^^^^^^^^^
28 |     pub extranonce2_len: usize,
   |         ^^^^^^^^^^^^^^^
   |
   = note: `SubmitShareWithChannelId` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
   = note: `#[warn(dead_code)]` on by default

warning: field `0` is never read
  --> translator/src/lib/error.rs:16:9
   |
15 |     SubmitSharesExtended(
   |     -------------------- field in this variant
16 |         async_channel::SendError<roles_logic_sv2::mining_sv2::SubmitSharesExtended<'a...
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
16 |         (),
   |         ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:18:20
   |
18 |     SetNewPrevHash(async_channel::SendError<roles_logic_sv2::mining_sv2::SetNewPrevHash<'a>...
   |     -------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
18 |     SetNewPrevHash(()),
   |                    ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:19:26
   |
19 |     NewExtendedMiningJob(async_channel::SendError<NewExtendedMiningJob<'a>...
   |     -------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
19 |     NewExtendedMiningJob(()),
   |                          ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:20:12
   |
20 |     Notify(tokio::sync::broadcast::error::SendError<Notify<'a>>),
   |     ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
20 |     Notify(()),
   |            ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:21:15
   |
21 |     V1Message(async_channel::SendError<v1::Message>),
   |     --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
21 |     V1Message(()),
   |               ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:22:13
   |
22 |     General(String),
   |     ------- ^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
22 |     General(()),
   |             ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:23:16
   |
23 |     Extranonce(async_channel::SendError<(ExtendedExtranonce, u32)>),
   |     ---------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |
   |     field in this variant
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
23 |     Extranonce(()),
   |                ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:25:9
   |
24 |     SetCustomMiningJob(
   |     ------------------ field in this variant
25 |         async_channel::SendError<roles_logic_sv2::mining_sv2::SetCustomMiningJob<'a...
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
25 |         (),
   |         ~~

warning: field `0` is never read
  --> translator/src/lib/error.rs:28:9
   |
27 |       NewTemplate(
   |       ----------- field in this variant
28 | /         async_channel::SendError<(
29 | |             roles_logic_sv2::template_distribution_sv2::SetNewPrevHash...
30 | |             Vec<u8>,
31 | |         )>,
   | |__________^
   |
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
28 |         (),
   |         ~~

warning: field `target` is never read
   --> translator/src/lib/proxy/bridge.rs:529:9
    |
525 | pub struct OpenSv1Downstream {
    |            ----------------- field in this struct
...
529 |     pub target: Arc<Mutex<Vec<u8>>>,
    |         ^^^^^^

warning: struct `TranslatorProxyUpstream` is never constructed
  --> translator/src/lib/proxy_config.rs:18:12
   |
18 | pub struct TranslatorProxyUpstream {
   |            ^^^^^^^^^^^^^^^^^^^^^^^

warning: struct `TranslatorProxyDownstream` is never constructed
  --> translator/src/lib/proxy_config.rs:25:12
   |
25 | pub struct TranslatorProxyDownstream {
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: associated function `new` is never used
  --> translator/src/lib/proxy_config.rs:32:12
   |
31 | impl ProxyConfig {
   | ---------------- associated function in this implementation
32 |     pub fn new(
   |            ^^^

warning: associated function `new` is never used
  --> translator/src/lib/proxy_config.rs:65:12
   |
64 | impl DownstreamDifficultyConfig {
   | ------------------------------- associated function in this implementation
65 |     pub fn new(
   |            ^^^

warning: fields `timestamp_of_last_update` and `should_aggregate` are never read
  --> translator/src/lib/proxy_config.rs:91:9
   |
87 | pub struct UpstreamDifficultyConfig {
   |            ------------------------ fields in this struct
...
91 |     pub timestamp_of_last_update: u64,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
92 |     #[serde(default = "bool::default")]
93 |     pub should_aggregate: bool,
   |         ^^^^^^^^^^^^^^^^
   |
   = note: `UpstreamDifficultyConfig` has derived impls for the traits `Clone` and `Debug`, but these are intentionally ignored during dead code analysis

warning: associated function `new` is never used
  --> translator/src/lib/proxy_config.rs:97:12
   |
96 | impl UpstreamDifficultyConfig {
   | ----------------------------- associated function in this implementation
97 |     pub fn new(
   |            ^^^

warning: struct `Sv2MiningConnection` is never constructed
  --> translator/src/lib/upstream_sv2/mod.rs:15:12
   |
15 | pub struct Sv2MiningConnection {
   |            ^^^^^^^^^^^^^^^^^^^
   |
   = note: `Sv2MiningConnection` has derived impls for the traits `Debug` and `Clone`, but these are intentionally ignored during dead code analysis

</code>
</details>

File: roles/translator/src/lib/proxy_config.rs
Comment (jbesraa) [reply roles/translator/src/lib/proxy_config.rs:53]: https://github.com/stratum-mining/stratum/issues/1116

**Comment by jbesraa** [comment]
> I tested the branch locally, and the logs for the translator look fine. However, I noticed quite a few warnings. I believe some of the error cases are no longer being triggered. If possible, could you remove those and resolve the other warnings as well? Everything else looks good.
> Warning Logs Translator
> `` fields `extranonce` and `extranonce2_len` are never read --> translator/src/lib/downstream_sv1/mod.rs:27:9 | 24 | pub struct SubmitShareWithChannelId { | ------------------------ fields in this struct ... 27 | pub extranonce: Vec, | ^^^^^^^^^^ 28 | pub extranonce2_len: usize, | ^^^^^^^^^^^^^^^ | = note: `SubmitShareWithChannelId` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis = note: `#[warn(dead_code)]` on by default warning: field `0` is never read --> translator/src/lib/error.rs:16:9 | 15 | SubmitSharesExtended( | -------------------- field in this variant 16 | async_channel::SendError<roles_logic_sv2::mining_sv2::SubmitSharesExtended<'a... | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 16 | (), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:18:20 | 18 | SetNewPrevHash(async_channel::SendError<roles_logic_sv2::mining_sv2::SetNewPrevHash<'a>... | -------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 18 | SetNewPrevHash(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:19:26 | 19 | NewExtendedMiningJob(async_channel::SendError<NewExtendedMiningJob<'a>... | -------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 19 | NewExtendedMiningJob(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:20:12 | 20 | Notify(tokio::sync::broadcast::error::SendError<Notify<'a>>), | ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 20 | Notify(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:21:15 | 21 | V1Message(async_channel::SendErrorv1::Message), | --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 21 | V1Message(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:22:13 | 22 | General(String), | ------- ^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 22 | General(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:23:16 | 23 | Extranonce(async_channel::SendError<(ExtendedExtranonce, u32)>), | ---------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | | | field in this variant | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 23 | Extranonce(()), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:25:9 | 24 | SetCustomMiningJob( | ------------------ field in this variant 25 | async_channel::SendError<roles_logic_sv2::mining_sv2::SetCustomMiningJob<'a... | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 25 | (), | ~~ warning: field `0` is never read --> translator/src/lib/error.rs:28:9 | 27 | NewTemplate( | ----------- field in this variant 28 | / async_channel::SendError<( 29 | | roles_logic_sv2::template_distribution_sv2::SetNewPrevHash... 30 | | Vec, 31 | | )>, | |__________^ | help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field | 28 | (), | ~~ warning: field `target` is never read --> translator/src/lib/proxy/bridge.rs:529:9 | 525 | pub struct OpenSv1Downstream { | ----------------- field in this struct ... 529 | pub target: Arc<Mutex<Vec>>, | ^^^^^^ warning: struct `TranslatorProxyUpstream` is never constructed --> translator/src/lib/proxy_config.rs:18:12 | 18 | pub struct TranslatorProxyUpstream { | ^^^^^^^^^^^^^^^^^^^^^^^ warning: struct `TranslatorProxyDownstream` is never constructed --> translator/src/lib/proxy_config.rs:25:12 | 25 | pub struct TranslatorProxyDownstream { | ^^^^^^^^^^^^^^^^^^^^^^^^^ warning: associated function `new` is never used --> translator/src/lib/proxy_config.rs:32:12 | 31 | impl ProxyConfig { | ---------------- associated function in this implementation 32 | pub fn new( | ^^^ warning: associated function `new` is never used --> translator/src/lib/proxy_config.rs:65:12 | 64 | impl DownstreamDifficultyConfig { | ------------------------------- associated function in this implementation 65 | pub fn new( | ^^^ warning: fields `timestamp_of_last_update` and `should_aggregate` are never read --> translator/src/lib/proxy_config.rs:91:9 | 87 | pub struct UpstreamDifficultyConfig { | ------------------------ fields in this struct ... 91 | pub timestamp_of_last_update: u64, | ^^^^^^^^^^^^^^^^^^^^^^^^ 92 | #[serde(default = "bool::default")] 93 | pub should_aggregate: bool, | ^^^^^^^^^^^^^^^^ | = note: `UpstreamDifficultyConfig` has derived impls for the traits `Clone` and `Debug`, but these are intentionally ignored during dead code analysis warning: associated function `new` is never used --> translator/src/lib/proxy_config.rs:97:12 | 96 | impl UpstreamDifficultyConfig { | ----------------------------- associated function in this implementation 97 | pub fn new( | ^^^ warning: struct `Sv2MiningConnection` is never constructed --> translator/src/lib/upstream_sv2/mod.rs:15:12 | 15 | pub struct Sv2MiningConnection { | ^^^^^^^^^^^^^^^^^^^ | = note: `Sv2MiningConnection` has derived impls for the traits `Debug` and `Clone`, but these are intentionally ignored during dead code analysis ``

Ill leave this to a different PR if thats ok?

File: roles/translator/src/lib/proxy_config.rs
Comment (plebhash) [reply roles/translator/src/lib/proxy_config.rs]: I relayed #1116 to https://github.com/stratum-mining/stratum/discussions/1069#discussioncomment-10397507

this is a good idea but probably not too urgent for now

File: roles/translator/src/main.rs
Comment (plebhash) [inline roles/translator/src/main.rs:125]: let's find an alternative to removing this

File: roles/translator/src/lib/proxy_config.rs
Comment (plebhash) [reply roles/translator/src/lib/proxy_config.rs]: also related: https://github.com/stratum-mining/stratum/discussions/1069#discussioncomment-10290212

File: roles/translator/src/lib/proxy_config.rs
Comment (plebhash) [inline roles/translator/src/lib/proxy_config.rs:18]: I feel this `struct` should be called `TranslatorProxyUpstreamConfig`

File: roles/translator/src/lib/proxy_config.rs
Comment (plebhash) [inline roles/translator/src/lib/proxy_config.rs:41]: same thing here... just call it `DonwstreamConfig`

File: roles/translator/src/lib/mod.rs
Comment (GitGab19) [inline roles/translator/src/lib/mod.rs:71]: Why did you remove `let mut interrupt_signal_future = Box::pin(tokio::signal::ctrl_c().fuse());` and its management?
I'm just comparing it to what we currently have on `dev` at https://github.com/stratum-mining/stratum/blob/0f0ee1e32a7cdeb9af83718364e0896b36744824/roles/translator/src/main.rs#L83

File: roles/translator/src/lib/mod.rs
Comment (GitGab19) [inline roles/translator/src/lib/mod.rs:270]: Where is this used?

File: roles/translator/src/lib/proxy_config.rs
Comment (GitGab19) [inline roles/translator/src/lib/proxy_config.rs:58]: Same warning here, it seems the new() is not used anywhere.
So this comment is related to the previous one of course

File: roles/translator/src/lib/proxy_config.rs
Comment (GitGab19) [inline roles/translator/src/lib/proxy_config.rs:104]: Same applies for the new() here

File: roles/translator/src/lib/proxy_config.rs
Comment (GitGab19) [inline roles/translator/src/lib/proxy_config.rs:136]: Same applies for the new() here

**Review by GitGab19 — COMMENTED** [review]
tACK.
But I left some comments, mostly due to warnings received while compiling it.

File: roles/translator/src/lib/mod.rs
Comment (plebhash) [reply roles/translator/src/lib/mod.rs:270]: here: https://github.com/jbesraa/stratum/blob/42b11af73c19bc1ba4158cccc22fbd668d7548e4/roles/tests-integration/tests/integration_e2e.rs#L35

File: roles/translator/src/lib/proxy_config.rs
Comment (plebhash) [reply roles/translator/src/lib/proxy_config.rs:58]: the `new` functiones are meant to be used in #1066 

should we suppress the warnings with a `#[allow(dead_code)]` or is there a more elegant solution?

Comment (plebhash) [reply roles/translator/src/lib/proxy_config.rs:58]: oh I see @jbesraa already addressed the comments around warnings

https://github.com/stratum-mining/stratum/pull/1092#issuecomment-2301771700

overall I'm not too opinionated on this... we could try to find a more elegant solution for a future PR, or just suppress warnings

Comment (plebhash) [reply roles/translator/src/lib/proxy_config.rs:58]: aside from this, I think this PR is pretty much good to go

Comment (GitGab19) [reply roles/translator/src/lib/proxy_config.rs:58]: > the `new` functiones are meant to be used in #1066
> 
> should we suppress the warnings with a `#[allow(dead_code)]` or is there a more elegant solution?
> 
> tbh `#[allow(dead_code)]` is fine for me, given that `roles` are just PoCs
> 
> we're getting similar warnings on #1095 and #1097, so if we go with this solution, we should bring it over to those PRs too

For me it's ok to leave it as it is. I just wanted to get more info on this, thank you!

Comment (jbesraa) [reply roles/translator/src/lib/proxy_config.rs:58]: fixed the warnings. its just a matter of making the unused structs public which I intended doing in #1066

File: roles/translator/src/lib/mod.rs
Comment (GitGab19) [reply roles/translator/src/lib/mod.rs:71]: @jbesraa do you have an answer to this? After that, for me it's ready to go

Comment (jbesraa) [reply roles/translator/src/lib/mod.rs:71]: sorry, missed this.
fixed

**Review by GitGab19 — APPROVED** [review]

**Review by plebhash — APPROVED** [review]

**Pull request merged**
